bootloader/piboot.go

```diff
@@ -0,0 +1,431 @@
+	reconfigBootloader := false
+	for k, v := range values {
+		// already set to the right value, nothing to do
+		if env.Get(k) == v {
+			continue
+		}
+		env.Set(k, v)
+		dirtyEnv = true
+		// Cases that change the bootloader configuration
+		if k == "snapd_recovery_mode" || k == "kernel_status" {
+			reconfigBootloader = true
+		}
+		if k == "snap_try_kernel" && v == "" {
+			// refresh (ok or not) finished, remove tryboot.txt
+			// (os_prefix in config.txt will be changed now in
+			// loadAndApplyConfig in the ok case)
+			trybootPath := filepath.Join(getSeedPartDir(), "tryboot.txt")
+			if err := os.Remove(trybootPath); err != nil {
+				logger.Noticef("cannot remove %s: %v", trybootPath, err)
+			}
+		}
+	}
+
+	if dirtyEnv {
+		if err := env.Save(); err != nil {
+			return err
+		}
+	}
+
+	if reconfigBootloader {
+		if err := p.loadAndApplyConfig(env); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *piboot) SetBootVarsFromInitramfs(values map[string]string) error {
+	env, err := ubootenv.OpenWithFlags(p.envFile(), ubootenv.OpenBestEffort)
+	if err != nil {
+		return err
+	}
+
+	dirtyEnv := false
+	for k, v := range values {
+		// already set to the right value, nothing to do
+		logger.Noticef("Setting %s=%s", k, v)
+		if env.Get(k) == v {
+			continue
+		}
+		env.Set(k, v)
+		dirtyEnv = true
+	}
+
+	if dirtyEnv {
+		if err := env.Save(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *piboot) loadAndApplyConfig(env *ubootenv.Env) error {
+	var prefix, cfgDir, dstDir string
+
+	cfgFile := "config.txt"
+	if env.Get("snapd_recovery_mode") == "run" {
+		kernelSnap := env.Get("snap_kernel")
+		kernStat := env.Get("kernel_status")
+		if kernStat == "try" {
+			// snap_try_kernel will be set when installing a new kernel
+			kernelSnap = env.Get("snap_try_kernel")
+			cfgFile = "tryboot.txt"
+		}
+		prefix = filepath.Join(pibootPartFolder, kernelSnap)
+		cfgDir = getSeedPartDir()
+		dstDir = filepath.Join(getSeedPartDir(), prefix)
+	} else {
+		// install/recovery modes, use recovery kernel
+		prefix = filepath.Join("/systems", env.Get("snapd_recovery_system"),
+			"kernel")
+		cfgDir = p.rootdir
+		dstDir = filepath.Join(p.rootdir, prefix)
+	}
+
+	logger.Debugf("configure piboot %s with prefix %q, cfgDir %q, dstDir %q",
+		cfgFile, prefix, cfgDir, dstDir)
+
+	if err := os.MkdirAll(dstDir, 0755); err != nil {
+		return err
+	}
+	return p.applyConfig(env, cfgFile, prefix, cfgDir, dstDir)
+}
+
+// Sets os_prefix in RPi config.txt or tryboot.txt
+func (p *piboot) setRPiCfgOsPrefix(prefix, inFile, outFile string) error {
+	buf, err := ioutil.ReadFile(inFile)
+	if err != nil {
+		return err
+	}
+
+	lines := strings.Split(string(buf), "\n")
+
+	replaced := false
+	newOsPrefix := "os_prefix=" + prefix + "/"
+	for i, line := range lines {
+		if strings.HasPrefix(line, "os_prefix=") {
+			lines[i] = newOsPrefix
+			replaced = true
+			break
+		}
+	}
```

Comment:
this should be robust against there being multiple os_prefix lines in the config.txt on the off chance that there are multiple ones, this function should handle that cleanly by deleting them both and writing only one of them so after this function runs there is only one os_prefix line in the file with the value we set it to

should probably also Noticef about this if it happens too

Reply:
Done now, also I've added some more unit tests